pages/a/orchestrating-crate-with-docker-112.html

```diff
@@ -0,0 +1,148 @@
+
+```bash
+$ eval $(docker-machine env crate-sw-1) # connect your local docker client to the remote machine
+$ docker service create \
+    --name crate \
+    --network crate-network \
+    --mode global \
+    --endpoint-mode dnsrr \
+    --update-parallelism 1 \
+    --update-delay 60s \
+    --mount type=bind,source=/tmp,target=/data \
+  crate:latest \
+    crate \
+    -Des.discovery.zen.ping.multicast.enabled=false \
+    -Des.discovery.zen.ping.unicast.hosts=crate \
+    -Des.gateway.expected_nodes=3 \
+    -Des.discovery.zen.minimum_master_nodes=2 \
+    -Des.gateway.recover_after_nodes=2 \
+    -Des.network.bind=_eth0:ipv4_
+```
+
+There are several things that need explanation in there:
+
+```bash
+    --name crate #\
+    --network crate-network \
+    --mode global \
+```
+These lines name the cluster 'crate', add the previously created overlay network to the service and set the scaling mode to "global", which means that there will always one container per node. Next the parameters:
+
+```bash
+    --endpoint-mode dnsrr \
+    --update-parallelism 1 \
+    --update-delay 60s \
+    --mount type=bind,source=/tmp,target=/data \
+```
+
+define the endpoint mode to be 'dnsrr' (DNS round robin - the dns entry 'crate' should cycle through the available IP addresses), `update-parallelism` and `update-delay` set how many instances should be updated in parallel and with what delay. Finally there is some volume mapping to ensure Crate is not writing data into the container (*don't use /tmp 's tmpfs for storing data. It will be gone after restarting*).
+
+```bash
+  crate:latest \
+    crate \
+    -Des.discovery.zen.ping.multicast.enabled=false \
+    -Des.discovery.zen.ping.unicast.hosts=crate \
+    -Des.gateway.expected_nodes=3 \
+    -Des.discovery.zen.minimum_master_nodes=2 \
+    -Des.gateway.recover_after_nodes=2 \
+    -Des.network.bind=_eth0:ipv4_
+```
+This part is the same for many Docker deployment of Crate: Multicast is disabled (it's not supported on the overlay network), instead the Docker DNS record 'crate' (the service name) is responsible for the cluster to find the other nodes. Since we are working with a 3-node cluster, the `expected_nodes`, `minimum_master_nodes`, and `recover_after_nodes` are set accordingly. The last line tells Crate to explicitly listen and bind to the IPv4 address of eth0 (usually the crate-network NIC), which is necessary since the auto discovery on the virtual adapters might be wrong.
+
+## Next Steps
+
+This setup gives you a working cluster but by setting the endpoint-mode to `DNSRR`, Docker disables its included load balancer and does not assign a public IP or port mapping to the cluster. The _other_ setting `VIP` (virtual IP) would provide a load-balanced public endpoint, but prohibits communication between the nodes. While we (and others) consider this a bug, a fix by Docker has yet (as of mid-September 2016) to be released.
```

Comment:
...working cluster, but because the endpoint-mode is set to 'DNSRR,' ...
also...is there some public discussion you can link to that proves your point about other people viewing this as a bug?